### PR TITLE
Address incorrect average temperature used for condensate volume calculation in DX and Water cooling coils

### LINKED ISSUE #9860: Incorrect average temperature used for condensate volume calculation in DXCoil object
Issue overview
--------------
Condensate volume calculation in DXCoil [here](https://github.com/NREL/EnergyPlus/blob/develop/src/EnergyPlus/DXCoils.cc#L14771) , CoilCoolingDX [here](https://github.com/NREL/EnergyPlus/blob/develop/src/EnergyPlus/Coils/CoilCoolingDX.cc#L733) and WaterCoil [here](https://github.com/NREL/EnergyPlus/blob/develop/src/EnergyPlus/WaterCoils.cc#L5199) uses incorrect average temperature. It should not be temperature difference, instead it should be absolute temperature. Only variable speed dx cooling coil uses correct equation [here](https://github.com/NREL/EnergyPlus/blob/develop/src/EnergyPlus/VariableSpeedCoils.cc#L6577-L6579).

```
current code:
Tavg = (state.dataDXCoils->DXCoil(DXCoilNum).InletAirTemp - state.dataDXCoils->DXCoil(DXCoilNum).OutletAirTemp) / 2.0;
```
```
Should have been:
Tavg = (state.dataDXCoils->DXCoil(DXCoilNum).InletAirTemp + state.dataDXCoils->DXCoil(DXCoilNum).OutletAirTemp) / 2.0;
```

### Details
Some additional details for this issue (if relevant):
 - Platform (any)
 - Version of EnergyPlus (develop, https://github.com/NREL/EnergyPlus/commit/3b44ce8acc08bb8895e3c7aa7eea6381b3081a2f)

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Defect file added (list location of defect file here)
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by Nigusse**
Pull request overview
---------------------
 - Fixes #9860
 - This fix corrects the average water temperature used to determine water density, which is then used to calculate the condensate volume flow rates. The report variables impacted are `Cooling Coil Condensate Volume Flow Rate` and `Cooling Coil Condensate Volume`. The change impacts `DXCoils`, `CoilCoolingDX`, and `WaterCoils` modules.

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [x] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [x] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [x] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [x] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [x] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: src/EnergyPlus/Coils/CoilCoolingDX.cc [hunk 1]
```diff
@@ -730,7 +730,7 @@ void CoilCoolingDX::simulate(EnergyPlus::EnergyPlusData &state,
         if (speedNum > 0) {
             // calculate and report condensation rates  (how much water extracted from the air stream)
             // water flow of water in m3/s for water system interactions
-            Real64 averageTemp = (evapInletNode.Temp - evapOutletNode.Temp) / 2.0;
+            Real64 averageTemp = (evapInletNode.Temp + evapOutletNode.Temp) / 2.0;
```
Comment (Nigusse) [inline src/EnergyPlus/Coils/CoilCoolingDX.cc:733]: corrected average water temperature calculation for evaluating water density.

File: src/EnergyPlus/DXCoils.cc [hunk 2]
```diff
@@ -14768,7 +14768,7 @@ void ReportDXCoil(EnergyPlusData &state, int const DXCoilNum) // number of the c
         // calculate and report condensation rates  (how much water extracted from the air stream)
         // water flow of water in m3/s for water system interactions
         //  put here to catch all types of DX coils
-        Tavg = (state.dataDXCoils->DXCoil(DXCoilNum).InletAirTemp - state.dataDXCoils->DXCoil(DXCoilNum).OutletAirTemp) / 2.0;
+        Tavg = (state.dataDXCoils->DXCoil(DXCoilNum).InletAirTemp + state.dataDXCoils->DXCoil(DXCoilNum).OutletAirTemp) / 2.0;
```
Comment (Nigusse) [inline src/EnergyPlus/DXCoils.cc]: corrected average water temperature calculation for evaluating water density.

File: src/EnergyPlus/WaterCoils.cc [hunk 3]
```diff
@@ -5136,87 +5135,66 @@ void ReportWaterCoil(EnergyPlusData &state, int const CoilNum)
...
-    Real64 RhoWater;
-    Real64 Tavg;
-    Real64 SpecHumOut;
-    Real64 SpecHumIn;
-    Real64 ReportingConstant;
 
-    if (state.dataWaterCoils->WaterCoil(CoilNum).reportCoilFinalSizes) {
+    auto &waterCoil = state.dataWaterCoils->WaterCoil(CoilNum);
+
+    if (waterCoil.reportCoilFinalSizes) {
```
Comment (Nigusse) [inline src/EnergyPlus/WaterCoils.cc:5141]: defined ShortCut for water coils as a good practice

File: src/EnergyPlus/WaterCoils.cc [hunk 4]
```diff
@@ -5136,87 +5135,66 @@ void ReportWaterCoil(EnergyPlusData &state, int const CoilNum)
...
         // water volumetric flow of water in m3/s for water system interactions
         //  put here to catch all types of DX coils
-        Tavg = (state.dataWaterCoils->WaterCoil(CoilNum).InletAirTemp - state.dataWaterCoils->WaterCoil(CoilNum).OutletAirTemp) / 2.0;
-
-        RhoWater = GetDensityGlycol(state,
-                                    state.dataPlnt->PlantLoop(state.dataWaterCoils->WaterCoil(CoilNum).WaterPlantLoc.loopNum).FluidName,
-                                    Tavg,
-                                    state.dataPlnt->PlantLoop(state.dataWaterCoils->WaterCoil(CoilNum).WaterPlantLoc.loopNum).FluidIndex,
-                                    RoutineName);
+        Real64 Tavg = (waterCoil.InletAirTemp + waterCoil.OutletAirTemp) / 2.0;
```
Comment (Nigusse) [inline src/EnergyPlus/WaterCoils.cc:5189]: corrected average water temperature calculation for evaluating water density.

File: src/EnergyPlus/WaterCoils.cc [hunk 5]
```diff
@@ -5136,87 +5135,66 @@ void ReportWaterCoil(EnergyPlusData &state, int const CoilNum)
...
-            state.dataWaterCoils->WaterCoil(CoilNum).CondensateVdot;
-        state.dataWaterData->WaterStorage(state.dataWaterCoils->WaterCoil(CoilNum).CondensateTankID)
-            .TwaterSupply(state.dataWaterCoils->WaterCoil(CoilNum).CondensateTankSupplyARRID) =
-            state.dataWaterCoils->WaterCoil(CoilNum).OutletAirTemp;
+        waterCoil.CondensateVdot =
+            max(0.0, (waterCoil.InletAirMassFlowRate * (waterCoil.InletAirHumRat - waterCoil.OutletAirHumRat) / Psychrometrics::RhoH2O(Tavg)));
+        waterCoil.CondensateVol = waterCoil.CondensateVdot * ReportingConstant;
+
+        state.dataWaterData->WaterStorage(waterCoil.CondensateTankID).VdotAvailSupply(waterCoil.CondensateTankSupplyARRID) = waterCoil.CondensateVdot;
+        state.dataWaterData->WaterStorage(waterCoil.CondensateTankID).TwaterSupply(waterCoil.CondensateTankSupplyARRID) = waterCoil.OutletAirTemp;
```
Comment (Nigusse) [inline src/EnergyPlus/WaterCoils.cc:5197]: applied shortcuts

File: tst/EnergyPlus/unit/DXCoils.unit.cc [hunk 6]
```diff
@@ -4330,6 +4348,21 @@ TEST_F(EnergyPlusFixture, MultiSpeedDXCoolingCoilOutputTest)
...
+    state->dataWaterData->WaterStorage(1).VdotAvailSupply.allocate(1);
+    state->dataWaterData->WaterStorage(1).TwaterSupply.allocate(1);
+    Coil.CondensateTankID = 1;
+    Coil.CondensateTankSupplyARRID = 1;
+    // calculate condensate vol flow rate
+    Real64 waterDensity = Psychrometrics::RhoH2O((Coil.InletAirTemp + Coil.OutletAirTemp) / 2.0);
+    Real64 results_condenstateVdot = Coil.InletAirMassFlowRate * (Coil.InletAirHumRat - Coil.OutletAirHumRat) / waterDensity;
+    ReportDXCoil(*state, DXCoilNum);
+    // check condensate volume flow rate
+    EXPECT_NEAR(results_condenstateVdot, Coil.CondensateVdot, 1.0E-11);
```
Comment (Nigusse) [inline tst/EnergyPlus/unit/DXCoils.unit.cc:4365]: Modified an existing multi-speed speed DX cooling coil to test condensation volume flow rate.

**Review by Nigusse — COMMENTED** [review]
Code change walk-through for PR #9863.

**Comment by Nigusse** [comment]
Comparison of the `develop` and `branch` report variable `Cooling Coil Condensate Volume`. The diffs are in the order of 1e-6 m3/s. 

image

**Comment by Nigusse** [comment]
Comparison of the `develop` and `branch` report variable `Cooling Coil Condensate Volume Flow Rate`. The diffs are in the order of 1e-8 m3/s.

image

**Comment by Nigusse** [comment]
This change impacts report variables `Cooling Coil Condensate Volume Flow Rate` and `Cooling Coil Condensate Volume`. Defect file has been uploaded to the `DefectFiles` folder on GitHub.

**Comment by Myoldmopar** [comment]
@Nigusse I'll take a look at this one now.

File: src/EnergyPlus/Coils/CoilCoolingDX.cc (hunk 1 above)
Comment (Myoldmopar) [reply src/EnergyPlus/Coils/CoilCoolingDX.cc:733]: 👍

File: src/EnergyPlus/DXCoils.cc [hunk 7]
```diff
@@ -14245,7 +14245,7 @@ void ReportDXCoil(EnergyPlusData &state, int const DXCoilNum) // number of the c
         // calculate and report condensation rates  (how much water extracted from the air stream)
         // water flow of water in m3/s for water system interactions
         //  put here to catch all types of DX coils
-        Real64 Tavg = (thisDXCoil.InletAirTemp - thisDXCoil.OutletAirTemp) / 2.0;
+        Real64 Tavg = (thisDXCoil.InletAirTemp + thisDXCoil.OutletAirTemp) / 2.0;
```
Comment (Myoldmopar) [inline src/EnergyPlus/DXCoils.cc:14248]: 👍

File: src/EnergyPlus/WaterCoils.cc (hunk 3 above)
Comment (Myoldmopar) [reply src/EnergyPlus/WaterCoils.cc:5141]: Thanks @Nigusse 👍

File: src/EnergyPlus/WaterCoils.cc (hunk 4 above)
Comment (Myoldmopar) [reply src/EnergyPlus/WaterCoils.cc:5189]: 👍

File: tst/EnergyPlus/unit/DXCoils.unit.cc [hunk 8]
```diff
@@ -4121,6 +4123,22 @@ TEST_F(EnergyPlusFixture, SingleSpeedDXCoolingCoilOutputTest)
...
+    state->dataWaterData->WaterStorage(1).TwaterSupply.allocate(1);
+    Coil.CondensateTankID = 1;
+    Coil.CondensateTankSupplyARRID = 1;
+    // calculate condensate vol flow rate
+    Real64 waterDensity = Psychrometrics::RhoH2O((Coil.InletAirTemp + Coil.OutletAirTemp) / 2.0);
+    Real64 results_condenstateVdot = Coil.InletAirMassFlowRate * (Coil.InletAirHumRat - Coil.OutletAirHumRat) / waterDensity;
+    Coil.DXCoilType = "Coil:Cooling:DX:SingleSpeed";
+    ReportDXCoil(*state, DXCoilNum);
+    // check condensate volume flow rate
+    EXPECT_NEAR(results_condenstateVdot, Coil.CondensateVdot, 1.0E-11);
```
Comment (Myoldmopar) [inline tst/EnergyPlus/unit/DXCoils.unit.cc:4141]: 👍

File: tst/EnergyPlus/unit/DXCoils.unit.cc (hunk 6 above)
Comment (Myoldmopar) [reply tst/EnergyPlus/unit/DXCoils.unit.cc:4365]: I'm glad you were able to reuse tests!

**Review by Myoldmopar — APPROVED** [review]
This looks great and sensible to me.  Building and testing locally now.

**Comment by Myoldmopar** [comment]
CI is happy, the diffs appear small, so they aren't even registering on our Decent comparisons.  Everything is happy with develop pulled in again.  Merging this, thanks @Nigusse

**Pull request merged**